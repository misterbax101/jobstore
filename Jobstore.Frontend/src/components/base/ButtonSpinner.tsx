import React from 'react';
import { Spinner } from 'reactstrap';

interface ButtonSpinnerProps {
    loading: boolean
}

const ButtonSpinner: React.FC<ButtonSpinnerProps> = ({ loading }) => {
    return loading ? <Spinner type="grow" color="secondary" className="align-middle" /> : null;
}
export default ButtonSpinner;
