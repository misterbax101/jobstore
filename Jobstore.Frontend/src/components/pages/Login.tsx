import React from 'react';
import { Formik, FormikProps, Form, Field, FormikActions } from 'formik';
import { FormGroup, Button, Label, Spinner } from 'reactstrap';
import { RouteComponentProps, Link } from 'react-router-dom';
import * as Yup from 'yup';

import Alert from '../../models/Alert';
import LoginModel from '../../models/LoginModel';
import CustomInput from '../base/CustomInput';
import AlertMessage from '../base/AlertMessage';

interface LoginProps extends RouteComponentProps {
    onLogin(model: LoginModel): Promise<void>;
    alert: Alert,
    isAuthenticated: boolean
}

const loginModelSchema = Yup.object().shape({
    email: Yup.string()
        .email('Invalid email')
        .required('Required'),
    password: Yup.string()
        .required('Required')
});


class Login extends React.Component<LoginProps, {}>{
    constructor(props: LoginProps) {
        super(props);
        if (props.isAuthenticated) {
            props.history.push('/');
        }
    }
    renderForm(filedProps: FormikProps<LoginModel>): JSX.Element {
        return (
            <Form>
                <FormGroup>
                    <Label htmlFor="email">Email</Label>
                    <Field type="email" name="email" placeholder="Enter email" component={CustomInput} />
                </FormGroup>
                <FormGroup>
                    <Label htmlFor="password">Password</Label>
                    <Field type="password" name="password" placeholder="Enter password" component={CustomInput} />
                </FormGroup>
                <FormGroup check>
                    <Label check>
                        <Field type="checkbox" name="remeberMe" component={CustomInput} />
                        Remeber me</Label>
                </FormGroup>
                <FormGroup>
                    <Button
                        color="primary"
                        type="submit"
                        disabled={filedProps.isSubmitting}>
                        Submit
                </Button>
                {filedProps.isSubmitting && <Spinner type="grow" color="secondary" style={{verticalAlign:'middle'}} />}
                    <Link to={'/sign-up'} className={'btn btn-link'}>Register</Link>
                </FormGroup>
            </Form>
        );
    }

    onFormSubmit = async (values: LoginModel, action: FormikActions<LoginModel>): Promise<void> => {
        await this.props.onLogin(values);
        action.setSubmitting(false);
    }

    render() {
        return (
            <React.Fragment>
                {this.props.alert && <AlertMessage data={this.props.alert} />}
                <Formik
                    initialValues={{ email: 'test@gmail.com', password: 'Aa!123456', remeberMe: false }}
                    validationSchema={loginModelSchema}
                    onSubmit={this.onFormSubmit}
                    render={this.renderForm}
                />
            </React.Fragment>
        );
    }
}

export default Login;