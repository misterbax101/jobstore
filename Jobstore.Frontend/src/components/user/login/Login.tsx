import React from 'react';
import { Formik, FormikProps, Form, Field, FormikActions } from 'formik';
import { FormGroup, Button, Container, Col, Label, Spinner, Alert } from 'reactstrap';
import { RouteComponentProps, Link } from 'react-router-dom';

import { LoginModel } from '../../../types';
import FormInput from '../../base/FormInput';
import CustomInput from '../../base/CustomInput';
import { loginValidationSchema } from './loginValidationSchema'
import resources from '../../../translations';
import { routes } from '../../../constants';
const { fields, placholders } = resources.common;

interface LoginProps extends RouteComponentProps {
    onLogin(model: LoginModel): Promise<boolean>;
    error: string | null,
    loading: boolean,
    isAuthenticated: boolean
}

class Login extends React.Component<LoginProps, {}>{
    constructor(props: LoginProps) {
        super(props);
        if (props.isAuthenticated) {
            this.redirect();;
        }
    }

    redirect = () => {
        const { history, location } = this.props;
        const redirectUrl = location.state && location.state.from
            ? location.state.from :
            routes.home;
        history.push(redirectUrl);
    }

    renderForm = (filedProps: FormikProps<LoginModel>): JSX.Element => {
        const { loading } = this.props;
        return (
            <Container>
                <Form>
                    <Col>
                        <FormInput name="email" type="email" label={fields.email} placeholder={placholders.emailPlacholder} />
                    </Col>
                    <Col>
                        <FormInput name="password" type="password" label={fields.password} placeholder={placholders.passwordPlacholder} />
                    </Col>
                    <Col>
                        <FormGroup check>
                            <Label check>
                                <Field type="checkbox" name="remeberMe" component={CustomInput} />
                                {resources.login.remeberMe}</Label>
                        </FormGroup>
                    </Col>
                    <Col>
                        <FormGroup>
                            <Button
                                color="primary"
                                type="submit"
                                disabled={loading}>
                                {resources.common.buttonLabels.submit}
                            </Button>
                            {loading && <Spinner type="grow" color="secondary" style={{ verticalAlign: 'middle' }} />}
                            <Link to={routes.signUp} className="btn btn-link">{resources.login.register}</Link>
                        </FormGroup>
                    </Col>
                </Form>
            </Container>
        );
    }

    onFormSubmit = async (values: LoginModel, action: FormikActions<LoginModel>): Promise<void> => {
        const result = await this.props.onLogin(values);
        action.setSubmitting(false);
        if (result) {
            this.redirect();
        }
    }

    render() {
        return (
            <React.Fragment>
                <Col md={{ size: 8, offset: 2 }}>
                    <h2>Login</h2>
                    {this.props.error && <Alert color='danger'>{this.props.error}</Alert>}
                    <Formik
                        initialValues={{ email: 'test@gmail.com', password: 'Aa!123456', remeberMe: false }}
                        validationSchema={loginValidationSchema}
                        onSubmit={this.onFormSubmit}
                        render={this.renderForm}
                    />
                </Col>
            </React.Fragment>
        );
    }
}

export default Login;