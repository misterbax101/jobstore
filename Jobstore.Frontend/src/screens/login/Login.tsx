import React from 'react';
import { Formik, FormikProps, Form, Field, FormikActions } from 'formik';
import { FormGroup, Button, Label, Spinner, Alert } from 'reactstrap';
import { RouteComponentProps, Link } from 'react-router-dom';

import { LoginModel } from '../../models';
import CustomInput from '../../components/base/CustomInput';
import { loginValidationSchema } from './loginValidationSchema'
import resources from '../../translations';

const { fields, placholders } = resources.common;

interface LoginProps extends RouteComponentProps {
    onLogin(model: LoginModel): Promise<void>;
    error: string | null,
    loading: boolean,
    isAuthenticated: boolean
}


class Login extends React.Component<LoginProps, {}>{
    constructor(props: LoginProps) {
        super(props);
        if (props.isAuthenticated) {
            props.history.push('/');
        }
    }

    renderForm = (filedProps: FormikProps<LoginModel>): JSX.Element => {
        return (
            <Form>
                <FormGroup>
                    <Label htmlFor="email">{fields.email}</Label>
                    <Field type="email" name="email" placeholder={placholders.emailPlacholder} component={CustomInput} />
                </FormGroup>
                <FormGroup>
                    <Label htmlFor="password">{fields.password}</Label>
                    <Field type="password" name="password" placeholder={placholders.passwordPlacholder} component={CustomInput} />
                </FormGroup>
                <FormGroup check>
                    <Label check>
                        <Field type="checkbox" name="remeberMe" component={CustomInput} />
                        {resources.login.remeberMe}</Label>
                </FormGroup>
                <FormGroup>
                    <Button
                        color="primary"
                        type="submit"
                        disabled={this.props.loading}>
                        {resources.common.buttonLabels.submit}
                </Button>
                    {this.props.loading && <Spinner type="grow" color="secondary" style={{ verticalAlign: 'middle' }} />}
                    <Link to={'/sign-up'} className={'btn btn-link'}>{resources.login.register}</Link>
                </FormGroup>
            </Form>
        );
    }

    onFormSubmit = async (values: LoginModel, action: FormikActions<LoginModel>): Promise<void> => {
        await this.props.onLogin(values);
        action.setSubmitting(false);
    }

    render() {
        return (
            <React.Fragment>
                {this.props.error && <Alert color='danger'>{this.props.error}</Alert>}
                <Formik
                    initialValues={{ email: 'test@gmail.com', password: 'Aa!123456', remeberMe: false }}
                    validationSchema={loginValidationSchema}
                    onSubmit={this.onFormSubmit}
                    render={this.renderForm}
                />
            </React.Fragment>
        );
    }
}

export default Login;