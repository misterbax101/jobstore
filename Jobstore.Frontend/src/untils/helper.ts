

export const truncateWithEllipses = (text: string, max: number = 100) => {
    return text.substr(0, max - 1) + (text.length > max ? '...' : '');
}


export const calculateSkip = (page: number, pageSize: number) => {
    return page == 0 ? 0 : (page - 1) * pageSize;
}


export const  getPagesCount =(totalCount: number, pageSize: number) =>  Math.ceil(totalCount / pageSize);