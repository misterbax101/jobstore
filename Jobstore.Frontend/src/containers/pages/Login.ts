import { connect } from 'react-redux';

import { Login } from './../../components/pages';
import { AppState } from './../../store';
import { login } from './../../store/auth/actions';

const mapStateToProps = (state: AppState) => ({
    isAuthenticated: state.auth.isAuthenticated,
    alertMessage: state.alert
});

export default connect(
    mapStateToProps,
    { onLogin: login },
)(Login)