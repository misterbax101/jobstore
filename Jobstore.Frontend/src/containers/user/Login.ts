import { connect } from 'react-redux';

import Login from '../../components/user/login/Login';
import { AppState } from '../../store';
import { login } from '../../store/auth/actions';

const mapStateToProps = (state: AppState) => ({
    isAuthenticated: state.auth.isAuthenticated,
    error: state.auth.error,
    loading: state.auth.loading
});

export default connect(
    mapStateToProps,
    { onLogin: login },
)(Login)