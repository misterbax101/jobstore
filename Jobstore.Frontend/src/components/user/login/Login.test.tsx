import React from 'react';
import { mount, shallow } from 'enzyme';
import { Alert, Button } from 'reactstrap';
import { Router } from 'react-router-dom';

import Login, { LoginProps } from './Login'
import ButtonSpinner from '../../base/ButtonSpinner';

import { routerTestProps } from './../../../untils/routerHelper';

const { history, location, match } = routerTestProps('/', {});

const defaultProps: LoginProps = {
    isAuthenticated: false,
    error: undefined,
    onLogin: () => { return new Promise<boolean>(() => true) },
    resetForm: () => { },
    loading: false,
    history: history,
    location: location,
    match: match
}

describe('<Login />', () => {
    it('should render', () => {

        let wrapper = shallow(<Login {...defaultProps} />);

        expect(wrapper.getElements()).toMatchSnapshot();
    });

    it('should calls reset', () => {

        const resetMock = jest.fn(() => { });
        mount(
            <Router history={history}>
                <Login {...defaultProps} resetForm={resetMock} />
            </Router>);

        expect(resetMock).toBeCalled();
    });

    it('should redirects to home', () => {

        const redirectUrl = '/test';
        location.state = { from: redirectUrl };

        shallow<Login>(<Login {...defaultProps}
            isAuthenticated={true} />);

        expect(history.location.pathname).toBe(redirectUrl);
    });

    it('should shows error', () => {

        const errorMsg = "test error";
        const wrapper = mount(
            <Router history={history}>
                <Login {...defaultProps}
                    error={errorMsg} />
            </Router>);

        const alertWrapper = wrapper.find(Alert).first();

        expect(alertWrapper).toBeDefined();
        expect(alertWrapper.text()).toBe(errorMsg);
    });

    it('should shows loader', () => {

        const wrapper = mount(
            <Router history={history}>
                <Login {...defaultProps}
                    loading={true} />
            </Router>);

        const spinnerWrapper = wrapper.find(ButtonSpinner).first();
        expect(spinnerWrapper.props().loading).toBeTruthy();
    });


    it('should shows validation errors', () => {

        const wrapper = mount(
            <Router history={history}>
                <Login {...defaultProps}
                    loading={true} />
            </Router>);

        wrapper
            .find(Button)
            .simulate('click');
        

    });

});