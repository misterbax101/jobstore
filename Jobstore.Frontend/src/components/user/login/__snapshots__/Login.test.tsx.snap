// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render 1`] = `
Array [
  <React.Fragment>
    <Col
      className="mt-4"
      md={
        Object {
          "offset": 2,
          "size": 8,
        }
      }
      sm={
        Object {
          "size": 12,
        }
      }
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
      <h2>
        Login
      </h2>
      <Formik
        enableReinitialize={false}
        initialValues={
          Object {
            "email": "",
            "password": "",
            "remeberMe": false,
          }
        }
        isInitialValid={false}
        onSubmit={[Function]}
        render={[Function]}
        validateOnBlur={true}
        validateOnChange={true}
        validationSchema={
          ObjectSchema {
            "_blacklist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "_conditions": Array [],
            "_defaultDefault": [Function],
            "_deps": Array [],
            "_excludedEdges": Array [],
            "_exclusive": Object {},
            "_mutate": undefined,
            "_nodes": Array [
              "password",
              "email",
            ],
            "_options": Object {
              "abortEarly": true,
              "recursive": true,
            },
            "_type": "object",
            "_typeError": [Function],
            "_whitelist": RefSet {
              "list": Set {},
              "refs": Map {},
            },
            "fields": Object {
              "email": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                  "undefined": false,
                },
                "_label": "Email",
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
              "password": StringSchema {
                "_blacklist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "_conditions": Array [],
                "_deps": Array [],
                "_exclusive": Object {
                  "required": true,
                },
                "_label": "Password",
                "_mutate": undefined,
                "_options": Object {
                  "abortEarly": true,
                  "recursive": true,
                },
                "_type": "string",
                "_typeError": [Function],
                "_whitelist": RefSet {
                  "list": Set {},
                  "refs": Map {},
                },
                "tests": Array [
                  [Function],
                ],
                "transforms": Array [
                  [Function],
                ],
              },
            },
            "tests": Array [],
            "transforms": Array [
              [Function],
            ],
          }
        }
      />
    </Col>
  </React.Fragment>,
]
`;
